### PR TITLE
DM: add worker manager

### PR DESCRIPTION
**Pull request by GMHDBJD**
- add worker manager to schedule workers

### CODE REVIEW

**Comment by GMHDBJD** [comment]
waiting for #259

**Review by nongfushanquan — APPROVED** [review]

**Comment by GMHDBJD** [comment]
cc @lance6716

File: jobmaster/dm/task_manager_test.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ const (
 func TestMain(m *testing.M) {
```
Comment (lance6716) [inline jobmaster/dm/task_manager_test.go:23]: I forget this function in previous PR. Does it act like SetUpSuit? Is this all your purpose?

File: jobmaster/dm/runtime/worker_status.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+const (
+	WorkerCreating WorkerStage = iota
+	WorkerOnline
+	WorkerFinished
+	WorkerOffline
+	// WorkerDestroying
+)
+
+type WorkerStatus struct {
+	Task  string
```
Comment (lance6716) [inline jobmaster/dm/runtime/worker_status.go:18]: name can be TaskID, as where we use it

File: jobmaster/dm/runtime/worker_status.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	ID    lib.WorkerID
+	Unit  lib.WorkerType
+	Stage WorkerStage
+}
+
+func (w *WorkerStatus) IsOffline() bool {
+	return w.Stage == WorkerOffline
+}
+
+func (w *WorkerStatus) IsExpected() bool {
```
Comment (lance6716) [inline jobmaster/dm/runtime/worker_status.go:28]: Expected is too ambiguous, can ytou explain it's meaning?

File: jobmaster/dm/worker_manager.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+		if ok {
+			runningWorker = worker.(runtime.WorkerStatus)
+			nextUnit = getNextUnit(persistentTask, runningWorker)
+		} else if nextUnit, err = wm.getCurrentUnit(ctx, persistentTask); err != nil {
+			log.L().Error("get current unit failed", zap.String("task", taskID), zap.Error(err))
+			recordError = err
+			continue
+		}
+
+		if ok && runningWorker.IsExpected() && nextUnit == runningWorker.Unit {
```
Comment (lance6716) [inline jobmaster/dm/worker_manager.go:170]: hard to read :joy:

File: jobmaster/dm/runtime/worker_status.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	WorkerFinished
+	WorkerOffline
+	// WorkerDestroying
+)
+
+type WorkerStatus struct {
+	Task  string
+	ID    lib.WorkerID
+	Unit  lib.WorkerType
+	Stage WorkerStage
```
Comment (lance6716) [inline jobmaster/dm/runtime/worker_status.go:21]: this is the expected stage or current stage?

Comment (GMHDBJD) [reply jobmaster/dm/runtime/worker_status.go]: worker has no expected stage

File: jobmaster/dm/task_manager_test.go (hunk 1 above)
Comment (GMHDBJD) [reply jobmaster/dm/task_manager_test.go]: change to suit

File: jobmaster/dm/task_manager_test.go [hunk 6]
```diff
@@ -102,62 +94,64 @@ func TestUpdateTaskStatus(t *testing.T) {
...
 
-	taskManager.OperateTask(context.Background(), Update, jobCfg, nil)
+	require.NoError(t.T(), taskManager.OperateTask(context.Background(), Update, jobCfg, nil))
 	state, err = jobStore.Get(context.Background())
-	require.NoError(t, err)
+	require.NoError(t.T(), err)
 	job = state.(*metadata.Job)
-	require.Equal(t, job.Tasks[source1].Stage, metadata.StageRunning)
+	require.Equal(t.T(), job.Tasks[source1].Stage, metadata.StageRunning)
 	// TODO: should it be paused?
```
Comment (lance6716) [inline jobmaster/dm/task_manager_test.go:143]: I'm neutral

File: jobmaster/dm/runtime/worker_status.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package runtime
+
+import (
+	"github.com/hanfei1991/microcosm/lib"
+)
+
+type WorkerStage int
```
Comment (lance6716) [inline jobmaster/dm/runtime/worker_status.go:7]: "WorkerStage will be aligned to TaskStage periodically"

File: jobmaster/dm/task_manager_test.go [hunk 8]
```diff
@@ -236,25 +229,24 @@ func TestCheckAndOperateTasks(t *testing.T) {
 	}
 	taskManager.UpdateTaskStatus(dumpStatus1)
 	taskManager.UpdateTaskStatus(dumpStatus2)
-	require.NoError(t, taskManager.checkAndOperateTasks(context.Background(), job))
+	require.NoError(t.T(), taskManager.checkAndOperateTasks(context.Background(), job))
```
Comment (lance6716) [inline jobmaster/dm/task_manager_test.go:232]: we can check mockAgent if it's received some call, to align the worker status

File: jobmaster/dm/task_manager.go [hunk 9]
```diff
@@ -73,10 +75,10 @@ func (tm *TaskManager) OperateTask(ctx context.Context, op OperateType, jobCfg *
...
 	switch op {
+	case Create, Update:
+		return tm.jobStore.Put(ctx, metadata.NewJob(jobCfg))
 	case Delete:
 		return tm.jobStore.Delete(ctx)
-	case Update:
-		return tm.jobStore.Put(ctx, metadata.NewJob(jobCfg))
 	case Resume:
 		stage = metadata.StageRunning
 	case Pause:
```
Comment (lance6716) [inline jobmaster/dm/task_manager.go:84]: line 94: 
> UpdateTaskStatus is called when receive task status from worker

What does worker mean, JobMaster?

File: jobmaster/dm/task_manager_test.go [hunk 10]
```diff
@@ -307,7 +299,7 @@ func TestTaskManager(t *testing.T) {
 	// mock check by interval
```
Comment (lance6716) [inline jobmaster/dm/task_manager_test.go:299]: line 296: if **task1 paused unexpected**, I think the worker status has changed to error and Paused, and the expected task stage is still Running. But my understanding is wrong with your code 🤔

File: jobmaster/dm/task_manager_test.go [hunk 11]
```diff
@@ -319,7 +311,7 @@ func TestTaskManager(t *testing.T) {
 	taskManager.OperateTask(ctx, Pause, nil, []string{source2})
 	mockAgent.SetResult([]error{agentError, agentError, agentError, nil})
 	// paused eventually
-	require.Eventually(t, func() bool {
+	require.Eventually(t.T(), func() bool {
 		mockAgent.Lock()
 		defer mockAgent.Unlock()
 		return len(mockAgent.results) == 0 && mockAgent.stages[source2] == metadata.StagePaused
```
Comment (lance6716) [inline jobmaster/dm/task_manager_test.go:317]: line 322
> task2 offline

task can't be offline, worker can

**Review by lance6716 — COMMENTED** [review]
I think I didn't understand WorkerStatus and TaskStatus, please update the code comment when addressing PR comment

File: jobmaster/dm/task_manager.go (hunk 9 above)
Comment (GMHDBJD) [reply jobmaster/dm/task_manager.go:84]: It is in OnWorkerOnline, which we get the status of the task and then updateTaskStatus

File: jobmaster/dm/task_manager_test.go (hunk 10 above)
Comment (GMHDBJD) [reply jobmaster/dm/task_manager_test.go:299]: There are only 4 status of worker(creating/online/offline/finished), it a worker paused by error, it's regarded as 'Online'

File: jobmaster/dm/task_manager_test.go (hunk 8 above)
Comment (GMHDBJD) [reply jobmaster/dm/task_manager_test.go:232]: It will no call mockAgent here. If it calls unexpected, mockAgent will panic in the test.

**Review by lance6716 — APPROVED** [review]

**Pull request merged**